### PR TITLE
[clang][codegen] Don't mark "int" TBAA on FP libcalls with indirect args

### PR DESCRIPTION
**Pull request by MacDue**
On some targets, an FP libcall with argument types such as long double
will be lowered to pass arguments indirectly via pointers. When this is
the case we should not mark the libcall with "int" TBAA as it may lead
to incorrect optimizations.

Currently, this can be seen for long doubles on x86_64-w64-mingw32. The
`load x86_fp80` after the call is (incorrectly) marked with "int" TBAA
(overwriting the previous metadata for "long double").

Nothing seems to break due to this currently as the metadata is being
incorrectly placed on the load and not the call. But if the metadata
is moved to the call (which this patch ensures), LLVM will optimize out
the setup for the arguments.

### CODE REVIEW

**Comment by MacDue** [comment]
Note the first commit (https://github.com/llvm/llvm-project/pull/108853/commits/6db9f6d56f0bbd56d017156f858eae68653fbd1b) shows a correctness issue with what's currently upstream as the `load x86_fp80` is incorrectly marked with "int" TBAA metadata (overwriting its metadata for long double).

**Comment by mstorsjo** [comment]
Thanks for the fix! Unfortunately I'm way out of my depth when it comes to actually reviewing this change, so hopefully you can find someone else to check/review it - but I appreciate you finding the root cause so quickly!

**Comment by efriedma-quic** [comment]
How does this interact with #107598?

**Comment by MacDue** [comment]
> How does this interact with #107598?

I think it's solving the same problem, but a different way (looking at the final LLVM function type rather than checking the ABI information).

**Comment by zahiraam** [comment]
> > How does this interact with #107598?
> 
>Though this also changes things to ensure when TBAA data is set, it's always set on the call.

Wasn't already doing that? (setting the TBAA on the call?).

**Comment by MacDue** [comment]
> > > How does this interact with #107598?
> > 
> > 
> > Though this also changes things to ensure when TBAA data is set, it's always set on the call.
> 
> Wasn't already doing that? (setting the TBAA on the call?).

It was setting it on `cast<llvm::Instruction>(Call.getScalarVal());` not the call (which you can get via an output on `EmitCall()`). At least in this case that meant it was putting the TBAA metadata on the `load x86_fp80` after the call. I'm not sure if there's other cases where something similar could happen.

**Comment by zahiraam** [comment]
> > > > How does this interact with #107598?
> > > 
> > > 
> > > Though this also changes things to ensure when TBAA data is set, it's always set on the call.
> > 
> > 
> > Wasn't already doing that? (setting the TBAA on the call?).
> 
> It was setting it on `cast<llvm::Instruction>(Call.getScalarVal())` not necessarily the call (which you can get via an output on `EmitCall()`). At least in this case that meant it was putting the TBAA metadata on the `load x86_fp80` after the call. I'm not sure if there's other cases where something similar could happen.



> > > > How does this interact with #107598?
> > > 
> > > 
> > > Though this also changes things to ensure when TBAA data is set, it's always set on the call.
> > 
> > 
> > Wasn't already doing that? (setting the TBAA on the call?).
> 
> It was setting it on `cast<llvm::Instruction>(Call.getScalarVal())` not necessarily the call (which you can get via an output on `EmitCall()`). At least in this case that meant it was putting the TBAA metadata on the `load x86_fp80` after the call. I'm not sure if there's other cases where something similar could happen.

Without this patch and without (https://github.com/llvm/llvm-project/pull/107598) the function `pow` doesn't generate `int TBAA` info on the call, but it does on a call to `cargl` with `-triple aarch64-unknown-unknown`.

`# |   %call = tail call fp128 @cargl([2 x fp128] noundef alignstack(16) undef) #1, !tbaa !2`

**Comment by MacDue** [comment]
> > > > > How does this interact with #107598?
> > > > 
> > > > 
> > > > Though this also changes things to ensure when TBAA data is set, it's always set on the call.
> > > 
> > > 
> > > Wasn't already doing that? (setting the TBAA on the call?).
> > 
> > 
> > It was setting it on `cast<llvm::Instruction>(Call.getScalarVal())` not necessarily the call (which you can get via an output on `EmitCall()`). At least in this case that meant it was putting the TBAA metadata on the `load x86_fp80` after the call. I'm not sure if there's other cases where something similar could happen.
> 
> > > > > How does this interact with #107598?
> > > > 
> > > > 
> > > > Though this also changes things to ensure when TBAA data is set, it's always set on the call.
> > > 
> > > 
> > > Wasn't already doing that? (setting the TBAA on the call?).
> > 
> > 
> > It was setting it on `cast<llvm::Instruction>(Call.getScalarVal())` not necessarily the call (which you can get via an output on `EmitCall()`). At least in this case that meant it was putting the TBAA metadata on the `load x86_fp80` after the call. I'm not sure if there's other cases where something similar could happen.
> 
> Without this patch and without (#107598) the function `pow` doesn't generate `int TBAA` info on the call, but it does on a call to `cargl` with `-triple aarch64-unknown-unknown`.
> 
> `# | %call = tail call fp128 @cargl([2 x fp128] noundef alignstack(16) undef) #1, !tbaa !2`

I'm not sue I follow, but the issue I spotted was "int" TBAA metadata was being set on the load following the `pow` call, but it has the same root cause as the `cargl` issue.

See the diff from the first commit:

```diff
diff --git a/clang/test/CodeGen/math-libcalls-tbaa-indirect-args.c b/clang/test/CodeGen/math-libcalls-tbaa-indirect-args.c
index dd013dcc8b3ca8..56c6b3ec00bc7e 100644
--- a/clang/test/CodeGen/math-libcalls-tbaa-indirect-args.c
+++ b/clang/test/CodeGen/math-libcalls-tbaa-indirect-args.c
@@ -19,7 +19,7 @@ long double powl(long double a, long double b);
 // CHECK-NEXT:    call void @llvm.lifetime.start.p0(i64 16, ptr nonnull [[BYVAL_TEMP1]]) #[[ATTR3]]
 // CHECK-NEXT:    store x86_fp80 0xK40008000000000000000, ptr [[BYVAL_TEMP1]], align 16, !tbaa [[TBAA3]]
 // CHECK-NEXT:    call void @powl(ptr dead_on_unwind nonnull writable sret(x86_fp80) align 16 [[TMP]], ptr noundef nonnull [[BYVAL_TEMP]], ptr noundef nonnull [[BYVAL_TEMP1]]) #[[ATTR3]]
-// CHECK-NEXT:    [[TMP0:%.*]] = load x86_fp80, ptr [[TMP]], align 16, !tbaa [[TBAA7:![0-9]+]]
+// CHECK-NEXT:    [[TMP0:%.*]] = load x86_fp80, ptr [[TMP]], align 16, !tbaa [[TBAA3]]
 // CHECK-NEXT:    call void @llvm.lifetime.end.p0(i64 16, ptr nonnull [[BYVAL_TEMP]]) #[[ATTR3]]
 // CHECK-NEXT:    call void @llvm.lifetime.end.p0(i64 16, ptr nonnull [[BYVAL_TEMP1]]) #[[ATTR3]]
 // CHECK-NEXT:    store x86_fp80 [[TMP0]], ptr [[AGG_RESULT]], align 16, !tbaa [[TBAA3]]
@@ -33,6 +33,4 @@ long double test_powl() {
 // CHECK: [[META4]] = !{!"long double", [[META5:![0-9]+]], i64 0}
 // CHECK: [[META5]] = !{!"omnipotent char", [[META6:![0-9]+]], i64 0}
 // CHECK: [[META6]] = !{!"Simple C/C++ TBAA"}
-// CHECK: [[TBAA7]] = !{[[META8:![0-9]+]], [[META8]], i64 0}
-// CHECK: [[META8]] = !{!"int", [[META5]], i64 0}
 //.
```

---

> but it does on a call to cargl with `-triple aarch64-unknown-unknown`.
`> `# |   %call = tail call fp128 @cargl([2 x fp128] noundef alignstack(16) undef) #1, !tbaa !2`

That looks okay though? It's not passing or returning values via pointers, so it should be safe to set the "int" TBAA (which indicates the only pointer it could read is errno).

**Comment by zahiraam** [comment]
> > > > > > How does this interact with #107598?
> > > > > 
> > > > > 
> > > > > Though this also changes things to ensure when TBAA data is set, it's always set on the call.
> > > > 
> > > > 
> > > > Wasn't already doing that? (setting the TBAA on the call?).
> > > 
> > > 
> > > It was setting it on `cast<llvm::Instruction>(Call.getScalarVal())` not necessarily the call (which you can get via an output on `EmitCall()`). At least in this case that meant it was putting the TBAA metadata on the `load x86_fp80` after the call. I'm not sure if there's other cases where something similar could happen.
> > 
> > 
> > > > > > How does this interact with #107598?
> > > > > 
> > > > > 
> > > > > Though this also changes things to ensure when TBAA data is set, it's always set on the call.
> > > > 
> > > > 
> > > > Wasn't already doing that? (setting the TBAA on the call?).
> > > 
> > > 
> > > It was setting it on `cast<llvm::Instruction>(Call.getScalarVal())` not necessarily the call (which you can get via an output on `EmitCall()`). At least in this case that meant it was putting the TBAA metadata on the `load x86_fp80` after the call. I'm not sure if there's other cases where something similar could happen.
> > 
> > 
> > Without this patch and without (#107598) the function `pow` doesn't generate `int TBAA` info on the call, but it does on a call to `cargl` with `-triple aarch64-unknown-unknown`.
> > `# | %call = tail call fp128 @cargl([2 x fp128] noundef alignstack(16) undef) #1, !tbaa !2`
> 
> I'm not sure I follow, but the issue I spotted was "int" TBAA metadata was being set on the load following the `pow` call, but it has the same root cause as the `cargl` issue.
> 
> See the diff from the first commit:
> 
> ```diff
> diff --git a/clang/test/CodeGen/math-libcalls-tbaa-indirect-args.c b/clang/test/CodeGen/math-libcalls-tbaa-indirect-args.c
> index dd013dcc8b3ca8..56c6b3ec00bc7e 100644
> --- a/clang/test/CodeGen/math-libcalls-tbaa-indirect-args.c
> +++ b/clang/test/CodeGen/math-libcalls-tbaa-indirect-args.c
> @@ -19,7 +19,7 @@ long double powl(long double a, long double b);
>  // CHECK-NEXT:    call void @llvm.lifetime.start.p0(i64 16, ptr nonnull [[BYVAL_TEMP1]]) #[[ATTR3]]
>  // CHECK-NEXT:    store x86_fp80 0xK40008000000000000000, ptr [[BYVAL_TEMP1]], align 16, !tbaa [[TBAA3]]
>  // CHECK-NEXT:    call void @powl(ptr dead_on_unwind nonnull writable sret(x86_fp80) align 16 [[TMP]], ptr noundef nonnull [[BYVAL_TEMP]], ptr noundef nonnull [[BYVAL_TEMP1]]) #[[ATTR3]]
> -// CHECK-NEXT:    [[TMP0:%.*]] = load x86_fp80, ptr [[TMP]], align 16, !tbaa [[TBAA7:![0-9]+]]
> +// CHECK-NEXT:    [[TMP0:%.*]] = load x86_fp80, ptr [[TMP]], align 16, !tbaa [[TBAA3]]
>  // CHECK-NEXT:    call void @llvm.lifetime.end.p0(i64 16, ptr nonnull [[BYVAL_TEMP]]) #[[ATTR3]]
>  // CHECK-NEXT:    call void @llvm.lifetime.end.p0(i64 16, ptr nonnull [[BYVAL_TEMP1]]) #[[ATTR3]]
>  // CHECK-NEXT:    store x86_fp80 [[TMP0]], ptr [[AGG_RESULT]], align 16, !tbaa [[TBAA3]]
> @@ -33,6 +33,4 @@ long double test_powl() {
>  // CHECK: [[META4]] = !{!"long double", [[META5:![0-9]+]], i64 0}
>  // CHECK: [[META5]] = !{!"omnipotent char", [[META6:![0-9]+]], i64 0}
>  // CHECK: [[META6]] = !{!"Simple C/C++ TBAA"}
> -// CHECK: [[TBAA7]] = !{[[META8:![0-9]+]], [[META8]], i64 0}
> -// CHECK: [[META8]] = !{!"int", [[META5]], i64 0}
>  //.
> ```
> 
> > but it does on a call to cargl with `-triple aarch64-unknown-unknown`.
> > `> `# |   %call = tail call fp128 @cargl([2 x fp128] noundef alignstack(16) undef) #1, !tbaa !2`
> 
> That looks okay though? It's not passing or returning values via pointers, so it should be safe to set the "int" TBAA (which indicates the only pointer it could read is errno).

Not really! `int TBAA` in in our downstream compiler is interpreted as describing the function arguments (they are not int) and the `load/store` of the argument before the library call are begin eliminated which result in unexpected behavior.
I am not objecting to anything; I am just wondering what difference it makes to have the TBAA attached to the call instead of how it is now.

**Comment by MacDue** [comment]
> Not really! `int TBAA` in in our downstream compiler is interpreted as describing the function arguments (they are not int) and the `load/store` of the argument before the library call are begin eliminated which result in unexpected behavior.
> I am not objecting to anything; I am just wondering what difference it makes to have the TBAA attached to the call instead of how it is now.

I believe the intention of both #96025 and #100302 was to set the metadata on the call. The case I mentioned where it was attached to the load after the call just seemed like an unintended consequence of attaching the metadata to the result value.

It sounds like both those changes break your downstream compiler?

**Comment by zahiraam** [comment]
> > Not really! int TBAA in in our downstream compiler is interpreted as describing the function arguments (they are not int) and the load/store of the argument before the library call are begin eliminated which result in unexpected behavior.
> 
> Oh wait, then you say "not really" you're only referring to the case where the arguments are passed _indirectly_ (i.e via pointers) not the case where they're passed via floating-point registers.
> 
Yes exactly! Have you tried running the test case in patch https://github.com/llvm/llvm-project/pull/107598 to see what IR it generates?

**Comment by MacDue** [comment]
> Yes exactly! Have you tried running the test case in patch #107598 to see what IR it generates?

I've updated my test (based on yours), and I think you can now see the TBAA metadata is only set on the call when the arguments are not being passed via pointers.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -690,23 +690,46 @@ static RValue emitLibraryCall(CodeGenFunction &CGF, const FunctionDecl *FD,
...
+    // arguments/results are passed directly. On some targets, types such as
+    // "long double" are passed indirectly via a pointer, and annotating the
+    // call with "int" TBAA metadata will lead to set up for those arguments
+    // being incorrectly optimized out.
+    bool ReturnAndAllArgumentsDirect =
+        isDirectOrIgnore(FnInfo->getReturnInfo()) &&
+        llvm::all_of(FnInfo->arguments(),
+                     [&](CGFunctionInfoArgInfo const &ArgInfo) {
+                       return isDirectOrIgnore(ArgInfo.info);
+                     });
```
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:722]: Make a predicate function and call at the end of the if expression below

Comment (MacDue) [reply clang/lib/CodeGen/CGBuiltin.cpp]: Done (and slightly tweaked the check)

**Comment by zahiraam** [comment]
OK. This LGTM. I guess we can adopt this PR instead of https://github.com/llvm/llvm-project/pull/107598 unless other reviewers have more to say.

**Review by zahiraam — APPROVED** [review]

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 2]
```diff
@@ -686,12 +686,31 @@ static Value *EmitSignBit(CodeGenFunction &CGF, Value *V) {
   return CGF.Builder.CreateICmpSLT(V, Zero);
 }
 
+/// Checks no arguments or results are passed indirectly in the ABI (i.e. via a
+/// hidden pointer). This is used to check annotating FP libcalls (that could
+/// set `errno`) with "int" TBAA metadata is safe. If any floating-point
+/// arguments are passed indirectly, setup for the call could be incorrectly
+/// optimized out.
+static bool HasNoIndirectArgumentsOrResults(CGFunctionInfo const &FnInfo) {
+  auto IsIndirect = [&](ABIArgInfo const &info) {
```
Comment (vfdff) [inline clang/lib/CodeGen/CGBuiltin.cpp:695]: May I ask why don't you determine whether the type of return and arguments is a pointer?  What's the difference?

Comment (MacDue) [reply clang/lib/CodeGen/CGBuiltin.cpp:695]: If you mean on the clang function declaration, they're not a pointer there (it's just how the ABI says to pass them).

If you mean why not check the LLVM function declaration, one reason is the pointers are opaque/untyped there, which may prevent setting the metadata on functions like `float frexpf(float, int*)` (which is a TODO in one of the tests).

**Review by vfdff — APPROVED** [review]
LGTM, thanks for help fix the issue

**Comment by MacDue** [comment]
I'll land this later today if there's no objections :slightly_smiling_face:

**Review by arsenm — APPROVED** [review]
LGTM, but like I mentioned on #107598, it would be good if there was a test that requires the argument check, and the return check isn't sufficient

**Review by arsenm — APPROVED** [review]

**Pull request merged**